spark/run_infer_vectors_service.py

```diff
@@ -0,0 +1,104 @@
+import requests
+import time
+import os
+import pyspark.sql.functions as F
+from pyspark.sql import SparkSession
+from pyspark.sql.types import StringType, ArrayType, FloatType
+from typing import List
+
+
+#DATASETS = ['arnetminer', 'kisti', 'pubmed', 'zbmath', 'inspire', 'qian', 'aminer']
+DATASETS = ['inspire']
+URL = 'https://rnd-rsbert-triton-gpu.literatumonline.com/v2/models/ensemble_model/infer'
+DELAY = 0.01
+
+s = requests.Session()
+
+def infer_vector(text: str) -> requests.Response:
+    request = {
+        "inputs": [
+            {
+                "name": "TEXT",
+                "shape": [1],
+                "datatype": "BYTES",
+                "data": [text]
+            }
+        ],
+        "outputs": [
+            {
+                "name": "sentence_emb",
+                "parameters": {"binary_data": False}
+            }
+        ]
+    }
+    reply = s.post(
+        url=URL,
+        json=request,
+        headers={'Content-Type': 'application/json'}
+    )
+    return reply
+
+@F.udf(returnType=ArrayType(elementType=FloatType()))
+def infer_vectors(text: str) -> List[float]:
+    """
+    Infer vectors for entries that do noy have vector repr
+    """
+    if text is None:
+        return None
+    while True:
+        # Sleep otherwise the service cannot handle the load
+        time.sleep(DELAY)
+        try:
+            reply = infer_vector(text=text)
+        except:
+            print('Connection exception, retrying...')
+            continue            
+        if reply.ok:
+            break
+        else:
+            print(f'Received status code: {reply.status_code}, retrying...')
+    vector = reply.json()['outputs'][0]['data']
+    return vector
+
+if __name__ == "__main__":
+    
+    spark = SparkSession.builder. \
+        appName('data_preprocessing'). \
+        config('spark.driver.memory', '58g'). \
+        config('spark.executor.cores', '16'). \
+        getOrCreate()
+
+    while True:
+        reply = infer_vector('test test test')
+        if reply.ok: break
```

Comment:
In the end, when we first the 200 response from the triton GPU, is ok to continue with the actual inference?

Reply:
Yes the service has waken up. But when the service autoscales, I get weird 502 error as reply, this is the reason i perform retries inside the udf.

Reply:
That's alright, good to know, thanks